File: FTAnalyser-RemoveFilters/Core/CensusFamily.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;

namespace FTAnalyzer
{
	public class CensusFamily : Family {

		private FactDate censusDate;
		private FactLocation bestLocation;

		public CensusFamily(Family f, FactDate censusDate) 
			: base(f)
		{
			this.censusDate = censusDate;
			this.bestLocation = null;
		}

		public bool process(FactDate censusDate, bool censusDone, bool includeResidence, bool lostCousinsCheck) {
			bool result = false;
			this.censusDate = censusDate;
            List<Fact> facts = new List<Fact>();
			if(isValidFamily()) {
				if (isValidIndividual(wife, censusDone, includeResidence, lostCousinsCheck, true)) {
					result = true;
					wife.Status = Individual.WIFE;
                    facts.AddRange(wife.AllFacts);
				} else 
					Wife = null;
				// overwrite bestLocation by husbands as most commonly the family
				// end up at husbands location after marriage
                if (isValidIndividual(husband, censusDone, includeResidence, lostCousinsCheck, true))
				{
					result = true;
					husband.Status = Individual.HUSBAND;
                    facts.AddRange(husband.AllFacts);
				} else 
					Husband = null;
				// update bestLocation by marriage date as husband and wife 
				// locations are often birth locations
				Fact marriage = getPreferredFact(Fact.MARRIAGE);
                if(marriage != null)
				    facts.Add(marriage);
				List<Individual> censusChildren = new List<Individual>();
				// sort children oldest first
				children.Sort(new CensusAgeComparator());
				foreach (Individual child in children) {
					// set location to childs birth location
					// this will end up setting birth location of last child 
					// as long as the location is at least Parish level
					child.Status = Individual.CHILD;
					if (isValidIndividual(child, censusDone, includeResidence, lostCousinsCheck, false))
					{
						result = true;
						censusChildren.Add(child);
                        facts.AddRange(child.AllFacts);
					}
				}
				children = censusChildren;
                this.bestLocation = FactLocation.BestLocation(facts, censusDate);
			}
			return result;
		}
		
		private bool isValidIndividual(Individual indiv, bool censusDone, bool includeResisdence, bool lostCousinsCheck, bool parentCheck) {
			if (indiv == null)
				return false;
			FamilyTree ft = FamilyTree.Instance;
			DateTime birth = indiv.BirthDate.StartDate;
			DateTime death = indiv.DeathDate.EndDate;
			if (birth < censusDate.StartDate && death > censusDate.StartDate && indiv.isCensusDone(censusDate, includeResisdence) == censusDone) {
				if(lostCousinsCheck && indiv.isLostCousinEntered(censusDate))
					return false;
				if (parentCheck) {
					// husband or wife with valid date range
					return true;
				} else {
					// individual is a child so remove if married before census date
					return !ft.isMarried(indiv, censusDate);
				}
			} else {
				return false;
			}
		}
		
		private bool isValidFamily() {
			if(MarriageDate.StartDate > censusDate.EndDate)
				return false;
			if (FamilyID == "IND")
				return true; // allow solo individual families to be processed
			// don't process family if either parent is under 16
			//if(husband != null) rtb.AppendText("husband : " + husband.getAge(censusDate) + "\n");
			if(husband != null && husband.getMaxAge(censusDate) < 16) 
				return false;
			//if(wife  != null) rtb.AppendText("wife : " + wife.getAge(censusDate) + "\n");
			if(wife != null && wife.getMaxAge(censusDate) < 16) 
				return false;
			return true;
		}
		
		public FactLocation BestLocation {
			get { return bestLocation; }
		}
		
		public int Relation {
			get
			{
				int relation = Individual.UNSET;
				foreach (Individual i in Members)
				{
					if (i.RelationType != Individual.UNKNOWN && i.RelationType < relation)
						relation = i.RelationType;
				}
				return relation == Individual.UNSET ? Individual.UNKNOWN : relation;
			}
		}
	}
}

File: Old Java Code/org/hardgate/familytree/core/CensusFamily.java
/*
 * Created on 15-Feb-2005
 *
 */
package org.hardgate.familytree.core;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.List;

import org.hardgate.familytree.comparators.CensusAgeComparator;
import org.hardgate.familytree.interfaces.FamilyLocal;

/**
 * @author A-Bisset
 *
 */
public class CensusFamily extends Family {

    private FactDate censusDate;
    private Location bestLocation;
    
    public CensusFamily (FamilyLocal fam) {
        super(fam);
    }
    
    public boolean process(FactDate censusDate) {
        boolean result = false;
        this.censusDate = censusDate;
        if(checkFamily()) {
	        this.bestLocation = updateBestLocation(new Location(), wife);
	        if (checkIndividual(wife)) {
			    result = true;
			    wife.setStatus(Individual.WIFE);
	        } else 
			    setWife(null);
		    // overwrite bestLocation by husbands as most commonly the family
		    // end up at husbands location after marriage
			this.bestLocation = updateBestLocation(bestLocation, husband);
			if (checkIndividual(husband)) {
			    result = true;
			    husband.setStatus(Individual.HUSBAND);
			} else 
			    setHusband(null);
			// update bestLocation by marriage date as husband and wife 
			// locations are often birth locations
			Fact marriage = getPreferredFact(Fact.MARRIAGE);
			this.bestLocation = updateBestLocation(bestLocation, marriage);
		    List<Individual> censusChildren = new ArrayList<Individual>();
			// sort children oldest first
			Collections.sort(children, new CensusAgeComparator());
			for (Individual child : children) {
			    // set location to childs birth location
			    // this will end up setting birth location of last child 
			    // as long as the location is at least Parish level
			    Fact birth = child.getPreferredFact(Fact.BIRTH);
			    this.bestLocation = updateBestLocation(bestLocation, birth);
			    child.setStatus(Individual.CHILD);
			    if (checkIndividual(child)) {
				    result = true;
				    censusChildren.add(child);
				}
			}
			children = censusChildren;
        }
		return result;
    }
    
    private Location updateBestLocation(Location bestLocation, Individual ind) {
        if (ind != null) {
		    Location location = ind.getBestLocation();
		    if (location.getLevel() >= Location.PARISH ||
		        location.getLevel() >= bestLocation.getLevel()) 
		        	return location;
        }
        return bestLocation;
    }
    
    private Location updateBestLocation(Location bestLocation, Fact fact) {
        if (fact != null) {
		    Location location = new Location(fact.getLocation());
		    if (location.getLevel() >= Location.PARISH ||
			    location.getLevel() >= bestLocation.getLevel()) 
		        	return location;
        }
        return bestLocation;
    }
    
    private boolean checkIndividual(Individual indiv) {
        if (indiv == null)
            return false;
        Client client = Client.getInstance();
        Calendar birth = (indiv.getBirthDate() == null) ? 
		        FactDate.MINDATE : indiv.getBirthDate().getStartDate();
		Calendar death = (indiv.getDeathDate() == null) ?
		        FactDate.MAXDATE : indiv.getDeathDate().getEndDate();
		if (birth.before(censusDate.getStartDate()) && 
		        death.after(censusDate.getStartDate()) && 
		        !indiv.isCensusDone(censusDate)) {
		    if (indiv.getStatus().equals(Individual.CHILD)) { 
		        // individual is a child so remove if married before census date
		    	return !client.isMarried(memberID, indiv, censusDate);
		    } else {
		        // husband or wife with valid date range
		        return true;
		    }
		} else {
		    return false;
		}
    }
    
    private boolean checkFamily() {
    	if(getMarriageDate().getStartDate().after(censusDate))
    		return false;
        // don't process family if either parent is under 16
    	//if(husband != null) System.out.println("husband : " + husband.getAge(censusDate));
    	if(husband != null && husband.getMaxAge(censusDate) < 16) 
    	    return false;
    	//if(wife  != null) System.out.println("wife : " + wife.getAge(censusDate));
    	if(wife != null && wife.getMaxAge(censusDate) < 16) 
    	    return false;
    	return true;
    }
    
    public Location getBestLocation() {
        return bestLocation;
    }
    
    public int getRelation() {
        int relation = Individual.UNSET;
        for(Individual i : getMembers()) {
            if (i.getRelation() != Individual.UNKNOWN &&
                i.getRelation() < relation) 
                	relation = i.getRelation();
        }
        return relation == Individual.UNSET ? Individual.UNKNOWN : relation;
    }
}
